### ISSUE TITLE
Listview inside PageView not updating data on Android properly, only on iOS

### ISSUE THREAD

**Issue by cetorres** [flutter/flutter#30936]
## Steps to Reproduce

iOS (works):
![Screen-Recording-2019-04-11-at-5 50 20-PM](https://user-images.githubusercontent.com/533661/55998054-aa111800-5c8a-11e9-9451-51c9b2914298.gif)

Android (doesn't work):
![Screen Recording 2019-04-11 at 5 52 00 PM mov](https://user-images.githubusercontent.com/533661/55999041-f3fbfd00-5c8e-11e9-88cb-9ca0ac4ccefb.gif)

## Logs

```
```

flutter analyze
```
No issues found! (ran in 2.7s)
```

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.2.1, on Mac OS X 10.14.3 18D109, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.2)
[✓] Android Studio (version 3.3)
[✓] VS Code (version 1.33.0)
[✓] Connected device (3 available)

• No issues found!
```

**Comment by najeira** [comment]
code for this?

**Comment by cetorres** [comment]
Let me explain a little more and show the code.

The “Todos” tab shows a list with all items. The “Caixas” tab shows only boxes from the all list (just as a shortcut for the user add boxes faster). The “Selecionados” tab shows the selected items (that the user added). When user changes on “Selecionados” tab a quantity of boxes, it’s not being reflected right the way on “Caixas” tab, on Android, only on iOS. Only after change to “Todos” tab and then change to “Caixas” tab that the quantities appear updated (on Android). What I’m doing to refresh the items is just reloading again the whole lists for “Todos” and “Caixas” tabs, with the quantities updated. The “Caixas” list is a subset of the “Todos” list, to make clear.

#### Load all items (Todos)
```dart
_loadRoomItems() async {
    RoomItemController controller = RoomItemController();
    RemoteData remoteData = await controller.getRoomItems(true);

    if (remoteData.error != null || remoteData.jsonString == null) {
      setState(() {
        _allItems = new List<Map>();
      });
      return;
    }

    List<RoomItem> roomItems = controller.convertToRoomItems(remoteData.jsonString);
    List allItems = roomItems.map((item) {
      return {"item": item, "quantity": 0};
    }).toList();

    setState(() {
      _allItems = allItems;
    });

    _loadBoxItems();
    _loadSurveyItems();

    // Schedule code to execute after widget on screen are loaded
    setState(() {
      SchedulerBinding.instance.addPostFrameCallback((_) => _afterFirstLayout(context));
    });
  }
```
#### Load boxes items (Caixas)
```dart
_loadBoxItems() async {
    setState(() {
      _boxItems = _allItems.where((item) {
        return (item["item"].name.toString().toLowerCase().contains('caixa') &&
            !item["item"].name.toString().toLowerCase().contains('som'));
      }).toList();
    });
  }
```

#### Set item quantity (when tap on minus and plus buttons)
```dart
_setSurveyItem(item, quantity) {
    setState(() {
      // Update all items
      _allItems = _allItems.map((map) {
        if (map["item"].id == item.id) {
          map["quantity"] = quantity;
        }
        return map;
      }).toList();

      // Update box items
      _loadBoxItems();

      // Update survey items and box items
      _surveyItems = List<SurveyItem>();
      _allItems.forEach((map) {
        if (map["quantity"] != null && map["quantity"] > 0) {
          _surveyItems.add(SurveyItem(item: map["item"], survey: _survey.id, quantity: map["quantity"]));
        }
      });
    });
  }
```

**Comment by najeira** [comment]
I can not figure out because there is only a part. Is there a minimal reproduction code?

**Comment by cetorres** [comment]
Here's the whole code for this screen.

```dart
import 'package:flutter/material.dart';
import 'package:imt_survey/controllers/room_item_controller.dart';
import 'package:imt_survey/controllers/survey_controller.dart';
import 'package:imt_survey/controllers/ws_controller.dart';
import 'package:imt_survey/models/room_item_model.dart';
import 'package:imt_survey/models/survey_item_model.dart';
import 'package:imt_survey/models/survey_model.dart';
import 'package:imt_survey/utils/util.dart';
import 'package:progress_dialog/progress_dialog.dart';
import '../utils/stepper.dart';
import 'package:flutter/scheduler.dart';

class SurveyItemsPage extends StatefulWidget {
  SurveyItemsPage({Key key, this.survey});

  final Survey survey;

  @override
  _SurveyItemsPageState createState() => _SurveyItemsPageState();
}

class _SurveyItemsPageState extends State<SurveyItemsPage> {
  Survey _survey;
  List<SurveyItem> _surveyItems;
  List<Map> _allItems;
  List<Map> _boxItems;
  int _selectedListIndex;
  PageController _pageController;
  ProgressDialog _progress;

  @override
  void initState() {
    super.initState();
    _survey = widget.survey;
    _selectedListIndex = 0;
    _pageController = PageController();
    _progress = new ProgressDialog(context);
    _progress.setMessage('Salvando...');
    _loadRoomItems();
  }

  // After page loads and finishes layout
  void _afterFirstLayout(BuildContext context) async {
    // If survey exists, change tab to selected itens
    if (_survey.id != null) {
      setState(() {
        _selectedListIndex = 2;
        _pageController.jumpToPage(_selectedListIndex);
      });
    }
  }

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  _loadRoomItems() async {
    RoomItemController controller = RoomItemController();
    RemoteData remoteData = await controller.getRoomItems(true);

    if (remoteData.error != null || remoteData.jsonString == null) {
      setState(() {
        _allItems = new List<Map>();
      });
      return;
    }

    List<RoomItem> roomItems = controller.convertToRoomItems(remoteData.jsonString);
    List allItems = roomItems.map((item) {
      return {"item": item, "quantity": 0};
    }).toList();

    setState(() {
      _allItems = allItems;
    });

    _loadBoxItems();
    _loadSurveyItems();

    // Schedule code to execute after widget on screen are loaded
    setState(() {
      SchedulerBinding.instance.addPostFrameCallback((_) => _afterFirstLayout(context));
    });
  }

  _loadBoxItems() async {
    setState(() {
      _boxItems = _allItems.where((item) {
        return (item["item"].name.toString().toLowerCase().contains('caixa') &&
            !item["item"].name.toString().toLowerCase().contains('som'));
      }).toList();
    });
  }

  _loadSurveyItems() async {
    if (_survey.id != null) {
      setState(() {
        // Update survey items
        _surveyItems = _survey.items;

        // Update all items
        _surveyItems.forEach((SurveyItem item) {
          _allItems = _allItems.map((allItem) {
            if (allItem["item"].id == item.item.id) {
              allItem["quantity"] = item.quantity;
            }
            return allItem;
          }).toList();
        });
      });
    }
  }

  _setSurveyItem(item, quantity) {
    setState(() {
      // Update all items
      _allItems = _allItems.map((map) {
        if (map["item"].id == item.id) {
          map["quantity"] = quantity;
        }
        return map;
      }).toList();

      // Update box items
      _loadBoxItems();

      // Update survey items and box items
      _surveyItems = List<SurveyItem>();
      _allItems.forEach((map) {
        if (map["quantity"] != null && map["quantity"] > 0) {
          _surveyItems.add(SurveyItem(item: map["item"], survey: _survey.id, quantity: map["quantity"]));
        }

      });
    });
  }

  _saveData() async {
    _progress.show();

    SurveyController controller = SurveyController();

    var response = await controller.saveSurvey(_survey);
    if (response['success'] == true) {
      Survey savedSurvey = Survey.fromJson(response['survey']);
      var items = _surveyItems.map((item) {
        item.survey = savedSurvey.id;
        return item;
      }).toList();
      var response2 = await controller.saveSurveyItems(savedSurvey.id, items);

      if (_progress.isShowing()) _progress.hide();

      if (response2['success'] == true) {
        Navigator.pop(context, true);
      } else {
        Util.showMessageDialog(context, 'Erro', 'Não foi possível salvar itens de vistoria.\n${response['error']}');
      }
    } else {
      if (_progress.isShowing()) _progress.hide();
      Util.showMessageDialog(context, 'Erro', 'Não foi possível salvar vistoria.\n${response['error']}');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(_survey.room.name),
        elevation: Theme.of(context).platform == TargetPlatform.iOS ? 0.0 : 4.0,
        actions: <Widget>[
          FlatButton(
            child: Text('Salvar', style: TextStyle(color: Colors.white, fontSize: 17)),
            onPressed: _saveData,
          ),
        ],
      ),
      body: (_allItems == null || _allItems.length == 0) ? Center(child: CircularProgressIndicator()) : _buildPages(),
      bottomNavigationBar: BottomNavigationBar(
        items: [
          BottomNavigationBarItem(
            title: Text('Todos'),
            icon: Icon(Icons.format_align_left),
          ),
          BottomNavigationBarItem(
            title: Text('Caixas'),
            icon: Icon(Icons.widgets),
          ),
          BottomNavigationBarItem(
            title: Text(_surveyItems == null || _surveyItems.length == 0
                ? 'Selecionados'
                : 'Selecionados (${_surveyItems.length})'),
            icon: Icon(Icons.done_all),
          ),
        ],
        currentIndex: _selectedListIndex,
        onTap: (index) {
          setState(() {
            _loadBoxItems();
            _selectedListIndex = index;
            _pageController.jumpToPage(index);
          });
        },
      ),
    );
  }

  PageView _buildPages() {
    return PageView(
      children: <Widget>[
        _buildAllList(),
        _buildBoxesList(),
        _buildSelectedList(),
      ],
      controller: _pageController,
      key: Key('PageView1'),
      onPageChanged: (index) {
        setState(() {
          _selectedListIndex = index;
        });
      },
    );
  }

  ListView _buildAllList() {
    return ListView.separated(
      key: new PageStorageKey('ListAll'),
      itemCount: _allItems.length,
      separatorBuilder: (BuildContext context, index) => Divider(),
      itemBuilder: (BuildContext context, index) {
        return ListTile(
          key: Key('AllTile_${_allItems[index]["item"].id}'),
          title: Text(_allItems[index]["item"].name),
          subtitle: Text('${_allItems[index]["item"].volume} m³'),
          trailing: Container(
            width: 110,
            child: SyStepper(
              key: Key("All_${_allItems[index]["item"].id}"),
              value: _allItems[index]["quantity"],
              size: 28,
              min: 0,
              onChange: (value) {
                _setSurveyItem(_allItems[index]["item"], value);
              },
            ),
          ),
        );
      },
    );
  }

  ListView _buildBoxesList() {
    return ListView.separated(
      key: new PageStorageKey('ListBoxes'),
      itemCount: _boxItems.length,
      separatorBuilder: (BuildContext context, index) => Divider(),
      itemBuilder: (BuildContext context, index) {
        return ListTile(
          key: Key('BoxesTile_${_boxItems[index]["item"].id}'),
          title: Text(_boxItems[index]["item"].name),
          subtitle: Text('${_boxItems[index]["item"].volume} m³'),
          trailing: Container(
            width: 110,
            child: SyStepper(
              key: Key("Boxes_${_boxItems[index]["item"].id}"),
              value: _boxItems[index]["quantity"],
              size: 28,
              min: 0,
              onChange: (value) {
                _setSurveyItem(_boxItems[index]["item"], value);
              },
            ),
          ),
        );
      },
    );
  }

  ListView _buildSelectedList() {
    return ListView.separated(
      key: new PageStorageKey('ListSelected'),
      itemCount: _surveyItems == null ? 0 : _surveyItems.length,
      separatorBuilder: (BuildContext context, index) => Divider(),
      itemBuilder: (BuildContext context, index) {
        return ListTile(
          key: Key('SelectedTile_${_surveyItems[index].item.id}'),
          title: Text(_surveyItems[index].item.name),
          subtitle: Text('${_surveyItems[index].item.volume} m³'),
          trailing: Container(
            width: 110,
            child: SyStepper(
              key: Key("Selected_${_surveyItems[index].item.id}"),
              value: _surveyItems[index].quantity,
              size: 28,
              min: 0,
              onChange: (value) {
                _setSurveyItem(_surveyItems[index].item, value);
              },
            ),
          ),
        );
      },
    );
  }
}

```

**Comment by cetorres** [comment]
On _setSurveyItem, _boxItems is being updated correctly inside the setState. But, for some reason, it's not being updated inside the ListView.

**Comment by cetorres** [comment]
I found the problem! It was the key on the SyStepper widget! I changed it to UniqueKey() on all three lists and it worked! Finally! It was driving me crazy! Because the state was being updated correctly! The list was updating, I tested showing the quantity inside the tile text. But the SyStepper widget was not being updated. When I changed the key to UniqueKey() it just worked! Great! Sorry, guys, if I made a mess with a bunch of code and confusion. But I was trying to pinpoint this problem for 1 week now. Thank you!
